src/catch2/internal/catch_sharding.hpp

```diff
@@ -0,0 +1,49 @@
+
+//              Copyright Catch2 Authors
+// Distributed under the Boost Software License, Version 1.0.
+//   (See accompanying file LICENSE_1_0.txt or copy at
+//        https://www.boost.org/LICENSE_1_0.txt)
+
+// SPDX-License-Identifier: BSL-1.0
+#ifndef CATCH_SHARDING_HPP_INCLUDED
+#define CATCH_SHARDING_HPP_INCLUDED
+
+#include <catch2/catch_session.hpp>
+
+#include <cmath>
+
+namespace Catch {
+
+    template<typename CONTAINER>
+    CONTAINER createShard(CONTAINER const& container, IConfig const& config) {
+        if (config.shardCount() > 1) {
+            size_t totalTestCount = container.size();
+
+            size_t shardCount = (std::min)(size_t(config.shardCount()), totalTestCount);
+            size_t shardIndex = (std::min)(size_t(config.shardIndex()), shardCount - 1);
+
+            double shardSize = totalTestCount / double(shardCount);
+
+            size_t startIndex = size_t(std::floor(shardIndex * shardSize));
+            size_t endIndex = size_t(std::floor((shardIndex + 1) * shardSize));
+
+            auto startIterator = std::next(container.begin(), startIndex);
+            auto endIterator = std::next(container.begin(), endIndex);
+
+            // Since we are calculating the end index with floating point numbers, but flooring 
+            // the value, we can't guarantee that the end index of the last shard lines up exactly
+            // with the end of input container. If we want the last shard, force the end index to
+            // be the end of the input container.
+            if (shardIndex == shardCount - 1) {
+                endIterator = container.end();
+            }
```

Comment:
I don't like the messing around with the floating point and then fixing away the potential error from floating point rounding.

Something like this should calculate the proper indices for a shard without running into these issues

```cpp
size_t testCount = container.size();
size_t shardCount = ...;

const size_t shardSize = testCount / shardCount;
const size_t leftoverTests = testCount % shardCount;

const size_t startIdx = shardIdx * shardSize + std::min(shardIdx, leftoverTests);
const size_t endIdx = (shardIdx + 1) * shardSize + std::min(shardIdx + 1, leftoverTests);
```

(this front-loads the extra tests, so that splitting 7 tests into 5 shards gives you shard of size [2, 2, 1, 1, 1])